### DISCUSSION TITLE
How does `supabase/realtime` work and what are tenants?

### DISCUSSION THREAD

**Discussion by Nikola-Milovic** [supabase/supabase#28668] [Questions] [answered]
Hello everyone, I get the gist of the features [supabase/realtime](https://github.com/supabase/realtime) offers, I am trying to self host it and use it for something different than it might be indented to. I (for now) want to use it only for listening to DB changes and sending appropriate notifications to my server (not users, but as single server).

What I am confused about is, what are the tenants? Is this a multitenant app, basically what supabase has? They have plenty of users each getting their own tenant for realtime? So if I am self hosting this, I would just create a single tenant? Would I even want more than one, any benefits to having more than 1?

Also for these channels, how are they related to tenants? Are they scoped per tenant? 

To summarize I am interested in what setup do I need to track DB changes in the self hosted instance of realtime and sending those notifications to our other server.

Secondly, if in the future I wanted to add realtime feature to my app how would I approach this? 

Both questions referring to tenancy

**Comment by GaryAustin1** [comment]
I can't answer your detailed questions, but realtime is multitenant in the Supabase infrastructure.  Supabase has realtime servers around the world and they are shared by Supabase instances.  So when a request is made to realtime it hits a realtime server and that server then makes a connection to the user's instance.

**Comment by johndpope** [comment, marked as answer]
**What is a Tenant in Supabase Realtime?**
Supabase Realtime supports multi-tenancy, allowing a single Realtime server to handle multiple independent databases (or "tenants"). Each tenant corresponds to a separate PostgreSQL database connection, with its own configuration for Change Data Capture (CDC), WebSocket channels, and security settings (like JWT secrets). Key points about tenants:

Tenant Configuration: Each tenant has a unique external_id, a database connection (host, user, password, etc.), and settings for CDC (e.g., publication, slot name).
Use Case: Multi-tenancy is useful if you’re running a single Realtime server that needs to serve multiple projects or databases (e.g., one Realtime server for tenant1 and tenant2, each with its own database).
Default Behavior: If you don’t register a tenant explicitly, Realtime can operate in a "single-tenant" mode, using the database connection specified in the .env file (e.g., POSTGRES_HOST, POSTGRES_DB). N.B. check the db name in docker-compose connection 
Do You Need to Register a Tenant?

Single Database: Your Terraform configuration and Supabase setup indicate you’re using a single RDS PostgreSQL database (mailopolyapp or postgres in the script). Your .env file likely specifies this database:
text



POSTGRES_HOST=bla.rds.amazonaws.com
POSTGRES_DB=blabla  # Or postgres, as in your script
POSTGRES_USER=mlypg_admin
POSTGRES_PASSWORD=<from_secrets_manager>
Default Tenant: Supabase Realtime automatically uses the database connection from .env as the default tenant if no tenants are explicitly registered via the /api/tenants endpoint. This is evident from your earlier steps, where we tested Realtime without registering a tenant, and you were able to create a publication (realtime) and a test table (test_realtime).
Conclusion: You do not need to register a tenant if you’re using a single database and the default configuration in .env is sufficient. The Realtime server is already connected to your RDS instance (as confirmed by your earlier steps), and it should be tracking changes in mailopolyapp (or postgres if that’s the database you’re using).

However, let’s explore why you might want to register a tenant and whether your script is correct for your setup.

When Should You Register a Tenant?
You might need to register a tenant if:

Multi-Tenancy: You plan to add more databases to the same Realtime server (e.g., a second RDS instance for a different project).
Custom Configuration: You want to override the default settings (e.g., different publication names, slot names, or polling intervals for specific databases).
Dynamic Management: You’re building a system where tenants are added programmatically via the API (e.g., for a SaaS application).

**Comment by Rubenburdin** [comment]
johndpope covered the tenant model well. Quick practical summary for self hosting:
Single database setup: You don't need to register tenants explicitly. The Realtime server uses your .env database connection as the default. Just configure POSTGRES_HOST, POSTGRES_DB, etc and it works.
For your use case (DB changes to your server): You have two options:

Postgres Changes via WebSocket: Connect your server as a client to Realtime and subscribe to changes. Works but means maintaining a persistent WS connection.
Database webhooks with pg_net: Probably simpler for server to server. Create triggers that POST to your endpoint on insert/update/delete. No need for Realtime at all:

`create trigger notify_server
  after insert or update on your_table
  for each row
  execute function supabase_functions.http_request(
    'https://your-server.com/webhook',
    'POST',
    '{"Content-Type":"application/json"}',
    '{}',
    '5000'
  );`

If you add user facing realtime later: Channels are scoped per tenant yes. Each client connects with a JWT, Realtime validates against that tenant's secret, and they subscribe to channels. You'd still just need one tenant for one database.
The tenancy model exists because Supabase runs shared Realtime infrastructure serving thousands of projects. For self hosting a single app, it's mostly irrelevant complexity you can ignore.
